packages/infra/src/handlers/process/lambda/prune.ts

```diff
@@ -0,0 +1,121 @@
+import {
+  type AliasConfiguration,
+  DeleteFunctionCommand,
+  type FunctionConfiguration,
+  LambdaClient,
+  ListAliasesCommand,
+  ListVersionsByFunctionCommand,
+} from '@aws-sdk/client-lambda';
+import { Env } from 'skuba-dive';
+
+import { getContext } from '../../framework/context';
+import { logger } from '../../framework/logging';
+
+import type { LambdaFunction } from './types';
+
+type Args = Pick<LambdaFunction, 'name'>;
+
+const lambda = new LambdaClient();
+
+export const prune = async (fns: Args[]): Promise<void> => {
+  await Promise.all(fns.map((fn) => pruneFunction(fn)));
+};
+
+export const pruneFunction = async ({ name }: Args): Promise<void> => {
+  const versionsToKeep = Env.nonNegativeInteger('VERSIONS_TO_KEEP');
+  const { abortSignal } = getContext();
+
+  const [aliases, versions] = await Promise.all([
+    listAliases(name, abortSignal),
+    listLambdaVersions(name, abortSignal),
+  ]);
+
+  const aliasMap = new Map(
+    aliases.flatMap((alias) =>
+      alias.FunctionVersion ? [[alias.FunctionVersion, alias]] : [],
+    ),
+  );
+
+  const versionsToPrune = versions
+    .filter(
+      (version) =>
+        version.Version &&
+        !aliasMap.has(version.Version) &&
+        version.Version !== '$LATEST',
+    )
+    .sort((a, b) => Number(b.Version) - Number(a.Version))
+    .slice(versionsToKeep);
+
+  if (!versionsToPrune.length) {
+    logger.info('No function versions to prune');
+    return;
+  }
+
+  logger.info(
+    `Pruning function versions: ${versionsToPrune
+      .map((version) => version.Version)
+      .join(', ')}`,
+  );
+
+  await Promise.all(
+    versionsToPrune.map((version) =>
+      lambda.send(
+        new DeleteFunctionCommand({
```

Comment:
Any potential concerns around this fanning out thousands of parallel requests on first run?

Could rate limiting / timeouts come into play? 😅

Reply:
Yeah I suppose the _deletes_ should be limited